// @ionic/cordova-builders/utils/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAngularAppConfig = exports.writeConfig = exports.readConfig = exports.addArchitectBuilder = exports.getDefaultAngularAppName = void 0;
const schematics_1 = require("@angular-devkit/schematics");
const CONFIG_PATH = 'angular.json';
function getDefaultAngularAppName(config) {
    const projects = config.projects;
    const projectNames = Object.keys(projects);
    for (const projectName of projectNames) {
        const projectConfig = projects[projectName];
        if (isAngularBrowserProject(projectConfig)) {
            return projectName;
        }
    }
    return projectNames[0];
}
exports.getDefaultAngularAppName = getDefaultAngularAppName;
function addArchitectBuilder(host, projectName, builderName, builderOpts) {
    const config = readConfig(host);
    const appConfig = getAngularAppConfig(config, projectName);
    appConfig.architect[builderName] = builderOpts;
    writeConfig(host, config);
}
exports.addArchitectBuilder = addArchitectBuilder;
function readConfig(host) {
    var _a;
    const sourceText = (_a = host.read(CONFIG_PATH)) === null || _a === void 0 ? void 0 : _a.toString('utf-8');
    if (!sourceText) {
        return;
    }
    return JSON.parse(sourceText);
}
exports.readConfig = readConfig;
function writeConfig(host, config) {
    host.overwrite(CONFIG_PATH, JSON.stringify(config, null, 2));
}
exports.writeConfig = writeConfig;
function isAngularBrowserProject(projectConfig) {
    if (projectConfig.projectType === 'application') {
        const buildConfig = projectConfig.architect.build;
        return buildConfig.builder === '@angular-devkit/build-angular:browser';
    }
    return false;
}
function getAngularAppConfig(config, projectName) {
    // eslint-disable-next-line no-prototype-builtins
    if (!config.projects.hasOwnProperty(projectName)) {
        throw new schematics_1.SchematicsException(`Could not find project: ${projectName}`);
    }
    const projectConfig = config.projects[projectName];
    if (isAngularBrowserProject(projectConfig)) {
        return projectConfig;
    }
    if (config.projectType !== 'application') {
        throw new schematics_1.SchematicsException(`Invalid projectType for ${projectName}: ${config.projectType}`);
    }
    else {
        const buildConfig = projectConfig.architect.build;
        throw new schematics_1.SchematicsException(`Invalid builder for ${projectName}: ${buildConfig.builder}`);
    }
}
exports.getAngularAppConfig = getAngularAppConfig;
